src/libYARP_OS/src/Node.cpp

```diff
@@ -213,122 +241,170 @@ class yarp::os::Node::Helper : public PortReader
 
     virtual bool read(ConnectionReader& reader);
 
-    Contact lookup(const ConstString& topic)
-    {
-        mutex.lock();
-        std::map<ConstString,NodeItem>::const_iterator i =
-            by_part_name.find(topic);
-        if (i == by_part_name.end()) {
-            mutex.unlock();
-            return Contact();
-        }
-        Contact c = i->second.contactable->where();
-        mutex.unlock();
-        return c;
-    }
-
     void getBusStats(NodeArgs& na)
     {
-        na.reply.addList();
+        na.reply = Value();
         na.success();
     }
 
     void getBusInfo(NodeArgs& na)
     {
-        na.reply.addList();
+        unsigned int opaque_id = 1;
+        ROSReport report;
+        Value v;
+        Bottle* connections = v.asList();
+
+        mutex.lock();
+        for (std::multimap<ConstString,NodeItem>::iterator it = by_part_name.begin(); it != by_part_name.end(); it++) {
+            NodeItem& item = it->second;
+            if (!(item.isSubscriber() || item.isPublisher())) {
+                continue;
+            }
+            item.update();
+            item.contactable->getReport(report);
+        }
+        mutex.unlock();
+
+        for (std::multimap<std::string, std::string>::const_iterator it = report.outgoingURIs.begin(); it != report.outgoingURIs.end(); ++it) {
+            Bottle& lst = connections->addList();
+            lst.addInt(opaque_id); // connectionId
+            lst.addString(it->second);
+            lst.addString("o");
+            lst.addString((it->second != "/yarp/node" ? "TCPROS" : "YARP"));
```

Comment:
@drdanz I think this needs a comment why there is the special case for `/yarp/node`. Not sure where this comes from.